src/Microsoft.AspNetCore.Razor.Language/Legacy/TokenizerBackedParser.cs

```diff
@@ -109,6 +109,53 @@ protected TSymbol Lookahead(int count)
             return symbols[count];
         }
 
+        /// <summary>
+        /// Looks forward until the specified condition is met.
+        /// </summary>
+        /// <param name="condition">A predicate accepting the symbol being evaluated and the list of symbols which have been looped through.</param>
+        /// <returns>true, if the condition was met. false - if the condition wasn't met and the last symbol has already been processed.</returns>
+        /// <remarks>The list of previous symbols is passed in the reverse order. So the last processed element will be the first one in the list.</remarks>
+        protected bool LookaheadUntil(Func<TSymbol, IEnumerable<TSymbol>, bool> condition)
+        {
+            if (condition == null)
+            {
+                throw new ArgumentNullException(nameof(condition));
+            }
+
+            bool matchFound = false;
+
+            // We add 1 in order to store the current symbol.
+            var symbols = new List<TSymbol>();
+            symbols.Add(CurrentSymbol);
+
+            while (true)
+            {
+                if (!NextToken())
+                {
+                    break;
+                }
+
+                symbols.Add(CurrentSymbol);
+                if (condition(CurrentSymbol, symbols.Reverse<TSymbol>()))
```

Comment:
Instead of constantly reversing the symbols i'd just `Insert(0)` when building the symbols list.

Reply:
That'll be way too slower. I just created a test to compare those two and for 100K operations current approach takes ~5ms, whereas with the proposed approach - ~700ms. Leaving as is.

Reply:
O_o super shocked. 👍